### ISSUE TITLE
clang-tidy: false positive readability-convert-member-functions-to-static to overloaded method

### ISSUE THREAD

**Issue by e-kwsm** [llvm/llvm-project#149152]
Prepare `/tmp/a.cpp`:

```cpp
struct S {
  void f();
  void f() const;

  int i = 0;
};

void S::f() { this->i++; }
void S::f() const { ; }
```

and `/tmp/compile_commands.json`:

```json
[{
  "directory": "/tmp",
  "command": "clang++ -c -o a.o a.cpp",
  "file": "a.cpp",
  "output": "a.o"
}]
```

```console
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 20.1.7
  Optimized build.
```

```console
$ clang-tidy --checks=readability-convert-member-functions-to-static --fix a.cpp
1 warning generated.
a.cpp:9:9: warning: method 'f' can be made static [readability-convert-member-functions-to-static]
    3 |   void f() const;
      |            ~~~~~
      |   static
    4 |
    5 |   int i = 0;
    6 | };
    7 |
    8 | void S::f() { this->i++; }
    9 | void S::f() const { ; }
      |         ^   ~~~~~
a.cpp:3:3: note: FIX-IT applied suggested code changes
    3 |   void f() const;
      |   ^
a.cpp:3:12: note: FIX-IT applied suggested code changes
    3 |   void f() const;
      |            ^
a.cpp:9:13: note: FIX-IT applied suggested code changes
    9 | void S::f() const { ; }
      |             ^
clang-tidy applied 3 of 3 suggested fixes.
```

gives the following invalid code

```cpp
struct S {
  void f();
  static void f() ;

  int i = 0;
};

void S::f() { this->i++; }
void S::f() { ; }
```

```console
$ clang++ -c -o a.o a.cpp
a.cpp:3:15: error: static and non-static member functions with the same parameter types cannot be overloaded
    3 |   static void f() ;
      |               ^
a.cpp:2:8: note: previous declaration is here
    2 |   void f();
      |        ^
a.cpp:9:9: error: redefinition of 'f'
    9 | void S::f() { ; }
      |         ^
a.cpp:8:9: note: previous definition is here
    8 | void S::f() { this->i++; }
      |         ^
2 errors generated.
```

<https://godbolt.org/z/aaYq98GMT>

**Comment by seanm** [comment]
I'm also seeing false positives with `readability-convert-member-functions-to-static`.  If I do what it says and make the thing `static` the compiler complains `error: invalid use of member 'ivar' in static member function`.

**Comment by gxyd** [comment]
Seems like a smaller MRE to reproduce the issue is (we don't need overloading of `f` to reproduce the MRE):
```c++
struct S {
  void f() const;
};

void S::f() const { ; }

```

[Godbolt link](https://godbolt.org/z/915r4Tr9b).

**Comment by gxyd** [comment]
Perhaps we can change the title accordingly, if one of the LLVM maintainer agrees as well?

**Comment by vbvictor** [comment]
> Seems like a smaller MRE to reproduce the issue is (we don't need overloading of `f` to reproduce the MRE):
> 
> struct S {
>   void f() const;
> };
> 
> void S::f() const { ; }
> [Godbolt link](https://godbolt.org/z/915r4Tr9b).

This don't seem like a bug. If you use --fix, it will remove const and add static, the code will be able to compile

**Comment by gxyd** [comment]
> > Seems like a smaller MRE to reproduce the issue is (we don't need overloading of `f` to reproduce the MRE):
> > struct S {
> > void f() const;
> > };
> > void S::f() const { ; }
> > [Godbolt link](https://godbolt.org/z/915r4Tr9b).
> 
> This don't seem like a bug. If you use --fix, it will remove const and add static, the code will be able to compile

I see.

I misunderstood, I thought since functionality-wise in general for a C++ program, one can't replace `void f() const` with `static void f()`, so the suggested change must be incorrect, which is where I was wrong, because `clang-tidy` only recommends this for the program under consideration (i.e. with special conditions like no use of `this` etc.).

Hence, resulting in me (incorrectly) believing that the *further reduced MRE* is the better MRE.

I'll keep this in mind, as I continue to work more on `clang-tidy`.

**Comment by gxyd** [comment]
@vbvictor , I'm thinking that the solution to this issue might possibly be that:

check whether two overloaded functions with exact same signatures barring `const`ness exist, then the `const` method shouldn't be converted to `static` usage.

Am I think along the correct line here? 

I might be able to use the logic (somewhat) from here: https://github.com/llvm/llvm-project/blob/9dd1eb49f74bdf0f0448ccff5f72b32b3f8fc9cc/clang-tools-extra/clang-tidy/bugprone/ReturnConstRefFromParameterCheck.cpp#L60-L77 .
